scripts/dwi_estimation_error_analysis.py

```diff
@@ -0,0 +1,642 @@
+    :obj:`~numpy.ndarray`
+        Training mask.
+
+    Raises
+    ------
+    ValueError
+        If the size of requested masked values is greater than available gradient vectors.
+
+    """
+    rng = np.random.default_rng(seed)
+
+    # Get the indices of the non-zero diffusion-encoding gradient vector indices
+    nnzero_degv_idx = np.where(~gtab.b0s_mask)[0]
+    if nnzero_degv_idx.size < size:
+        raise ValueError(
+            f"Requested {size} values for masking; gradient table has {nnzero_degv_idx.size} "
+            "non-zero diffusion-encoding gradient vectors. Reduce the number of masked values."
+        )
+    lo = rng.choice(nnzero_degv_idx, size=size, replace=False)
+
+    # Exclude the b0s
+    zero_degv_idx = np.asarray(list(set(range(len(gtab.bvals))).difference(nnzero_degv_idx)))
+    lo = np.hstack([zero_degv_idx, lo])
+    train_mask = np.ones(len(gtab.bvals), dtype=bool)
+    train_mask[lo] = False
+    return train_mask
+
+
+def perform_experiment(
+    gtab: gradient_table,
+    S0: float,
+    evals1: np.ndarray,
+    evecs: np.ndarray,
+    snr: float,
+    repeats: int,
+    kfold: list[int],
+) -> dict[int, list[tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]]]:
+    """
+    Perform the experiment by estimating the dMRI signal using a Gaussian process model.
+
+    Parameters
+    ----------
+    gtab : :obj:`~dipy.core.gradients.gradient_table`
+        Gradient table.
+    S0 : :obj:`float`
+        S0 value.
+    evals1 : :obj:`~numpy.ndarray`
+        Eigenvalues of the tensor.
+    evecs : :obj:`~numpy.ndarray`
+        Eigenvectors of the tensor.
+    snr : :obj:`float`
+        Signal-to-noise ratio.
+    repeats : :obj:`int`
+        Number of repeats.
+    kfold : :obj:`list`
+        list of directions to leave out/predict.
+
+    Returns
+    -------
+    :obj:`dict`
+        Data for the predicted signal and its error.
+
+    """
+
+    # Fix the random number generator for reproducibility when generating the
+    # signal
+    rng = np.random.default_rng(1234)
+
+    # Define the Gaussian process model parameter
+    lambda_s = 2.0
+    a = 1.0
+    sigma_sq = 0.5
+
+    data = {}
+    nzero_bvecs = gtab.bvecs[~gtab.b0s_mask]
+
+    # Simulate the fitting a number of times: every time the signal created will be a little
+    # different
+    # for _ in range(repeats):
+    # Create the DWI signal using a single tensor
+    signal = single_tensor(gtab, S0=S0, evals=evals1, evecs=evecs, snr=snr, rng=rng)
+
+    import pdb
+
+    pdb.set_trace()
+
+    # Loop over the number of indices that are left out from the training/need to be predicted
+    for n in kfold:
+        # Assumptions:
+        #  - Consecutive indices in the folds
+        #  - A single b0
+        kf = KFold(n_splits=n, shuffle=False)
+
+        # Define the Gaussian process model instance
+        gp_model = EddyMotionGPR(
+            kernel=SphericalKriging(a=a, lambda_s=lambda_s),
+            alpha=sigma_sq,
```

Comment:
Our $\sigma^2$ corresponds to $\alpha$ in Scikit-learn's implementation by comparison of Andersson 2015 and Rasmussen 2006 (https://gaussianprocess.org/gpml/chapters/RW.pdf).